### PR TITLE
Add option to skip deprecated field when recursively generate class spec

### PR DESCRIPTION
**Pull request by imagiczhang**

### CODE REVIEW

File: generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java [hunk 1]
```diff
@@ -70,6 +70,7 @@
   private static final String ARRAY_SUFFIX = "Array";
   private static final String MAP_SUFFIX = "Map";
   private static final String UNION_SUFFIX = "Union";
+  private static final String DEPRECATED = "deprecated";
```
[the remark this answers is not in the document]
Comment (imagiczhang) [reply generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java]: changed to use the existing constant.

File: generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java [hunk 2]
```diff
@@ -744,14 +754,17 @@ private RecordTemplateSpec generateRecord(RecordDataSchema schema)
...
-      processSchema(includedSchema, null, null);
+      processSchema(includedSchema, null, null, skipDeprecatedField);
     }
 
     final Map<CustomInfoSpec, Object> customInfoMap = new IdentityHashMap<>(schema.getFields().size() * 2);
 
     for (RecordDataSchema.Field field : schema.getFields())
     {
-      final ClassTemplateSpec fieldClass = processSchema(field.getType(), recordClass, field.getName());
+      if (skipDeprecatedField && isDeprecated(field)) {
```
Comment (evanw555) [inline generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java:764]: Leave a comment here explaining that this affects generation recursively (i.e. it skips generating this field, its type, and types referenced within this type). It's not totally clear just looking at it.

File: generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java [hunk 3]
```diff
@@ -131,9 +132,18 @@ public void registerDefinedSchema(DataSchema schema)
...
   public ClassTemplateSpec generate(DataSchema schema, DataSchemaLocation location)
+  {
+    return generate(schema, location, false);
+  }
+
+  /**
+   * Generate {@link ClassTemplateSpec} from the specified {@link DataSchema} and its location. If skipDeprecatedField
+   * is set to true, the class of the deprecated field will be bypassed in the generated specs.
+   */
+  public ClassTemplateSpec generate(DataSchema schema, DataSchemaLocation location, boolean skipDeprecatedField)
```
[the remark this answers is not in the document]
Comment (evanw555) [reply generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java:143]: Makes sense.

File: generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java [hunk 4]
```diff
@@ -131,9 +132,18 @@ public void registerDefinedSchema(DataSchema schema)
    * Generate {@link ClassTemplateSpec} from the specified {@link DataSchema} and its location.
    */
   public ClassTemplateSpec generate(DataSchema schema, DataSchemaLocation location)
+  {
+    return generate(schema, location, false);
+  }
+
+  /**
+   * Generate {@link ClassTemplateSpec} from the specified {@link DataSchema} and its location. If skipDeprecatedField
+   * is set to true, the class of the deprecated field will be bypassed in the generated specs.
```
Comment (evanw555) [inline generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java:141]: Change this to a proper `@param` Javadoc explanation, and please explain that the effect is recursive.

File: generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java (hunk 2 above)
Comment (imagiczhang) [reply generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java:767]: updated.

File: generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java (hunk 4 above)
Comment (imagiczhang) [reply generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java]: updated

**Review by evanw555 — APPROVED** [review]

**Comment by junchuanwang** [comment]
Please be mindful that this is open repo and please fill in description  @imagiczhang

**Pull request merged**